### PR TITLE
Edit variant notes in line

### PR DESCRIPTION
**Pull request by lizzij**
Clicking the edit button to toggle the note from a display-only view to an edit view in the page.

### CODE REVIEW

File: ui/shared/components/form/ReduxFormWrapper.jsx [hunk 1]
```diff
@@ -268,7 +271,11 @@ class ReduxFormWrapper extends React.Component {
   }
 
   componentWillUpdate(nextProps) {
-    if (nextProps.submitSucceeded && nextProps.closeOnSuccess && !nextProps.noModal) {
+    if (nextProps.submitSucceeded && nextProps.closeOnSuccess && nextProps.noModal) {
+      this.props.closeInlineModal()
+      this.props.handleClose(true)
```
Comment (hanars) [inline ui/shared/components/form/ReduxFormWrapper.jsx]: Before, if the `noModal` prop was passed `this.props.handleClose(true)` was NOT called, so this changes the behavior of every other time this component is rendered with that prop, almost definitely in a breaking way. Additionally, those forms do not have `closeInlineModal` passed in so calling it as a function will also break them. Please update this so it works for your components with the new prop without breaking backwards compatibility

File: ui/shared/components/form/ReduxFormWrapper.jsx [hunk 2]
```diff
@@ -147,6 +147,9 @@ class ReduxFormWrapper extends React.Component {
     /* React child component class. Mutually exclusive with fields */
     children: PropTypes.node,
 
+    /* Call if submit succeeded */
+    closeInlineModal: PropTypes.func,
```
Comment (hanars) [inline ui/shared/components/form/ReduxFormWrapper.jsx]: I would name this something more generic, like `onSubmitSucceeded`, so that you can pass any function you want to call after a successful submission

File: ui/shared/components/buttons/UpdateButton.jsx [hunk 3]
```diff
@@ -8,35 +8,39 @@ import Modal from '../modal/Modal'
...
-        initialValues={initialValues}
-        fields={formFields}
-        showErrorPanel={showErrorPanel}
-        confirmDialog={confirmDialog}
-        submitButtonText={submitButtonText}
-        confirmCloseIfNotSaved
-      />
-    ) }) }
-  </Modal>
+) => {
```
Comment (hanars) [inline ui/shared/components/buttons/UpdateButton.jsx]: It looks like nothing changed here functionally, so can you change this back to how it was?

File: ui/shared/components/panel/view-fields/BaseFieldView.jsx [hunk 4]
```diff
@@ -18,76 +20,126 @@ const FieldValue = styled.div`
...
+    const hasValue = (fieldValue && (!Object.getOwnPropertyNames(fieldValue).includes('length') || fieldValue.length > 0)) || this.props.showEmptyValues
+    if (!this.props.isEditable && !hasValue) {
+      return null
+    }
+    const fieldId = this.props.initialValues[this.props.idField]
+    const modalId = this.props.isEditable ? `edit-${fieldId || 'new'}-${this.props.field}-${this.props.modalId}` : null
+
+    const updateButton = this.props.showInLine ?
+      <div>
+        {this.state.showInLineButton &&
```
Comment (hanars) [inline ui/shared/components/panel/view-fields/BaseFieldView.jsx]: rather than doing a conditional for `this.state.showInLineButton` followed immediately by another conditional for `!this.state.showInLineButton` you could just use a ternary:
```
{this.state.showInLineButton ?
  <ButtonLink  icon="write"  onClick={this.toggleButtonVisibility}  /> : 
  <Segment>
          <ReduxFormWrapper
            noModal
            onSubmit={this.props.onSubmit}
            closeInlineModal={this.toggleButtonVisibility}
            form={this.props.modalId}
            initialValues={this.props.initialValues}
            fields={this.props.formFields}
            showErrorPanel={this.props.showErrorPanel}
            confirmDialog={this.props.confirmDialog}
            confirmCloseIfNotSaved
          />
  </Segment>
}
```

File: ui/shared/components/panel/view-fields/BaseFieldView.jsx [hunk 5]
```diff
@@ -18,76 +20,126 @@ const FieldValue = styled.div`
...
+          initialValues={this.props.initialValues}
+          formFields={this.props.formFields}
+          formContainer={<div style={this.props.modalStyle} />}
+          showErrorPanel={this.props.showErrorPanel}
+          confirmDialog={this.props.addConfirm}
+          size="tiny"
+        />
+      )
+
+    const editButton = this.props.isEditable && (this.props.formFields ?
```
Comment (hanars) [inline ui/shared/components/panel/view-fields/BaseFieldView.jsx]: So this is very nitpicky, but the way the control flow here is kind of confusing, because before if it wasn't editable `updateButton` was just never defined, and now you are always defining it and then sometimes just not using it. So for non-editable fields, or fields without `formFields` this is kind of confusing because there are all these unused components being made and its kind of hard to read through and follow the logic for what gets created.  I think rather than defining `updateButton` and then maybe assigning it to `editButton`, you should just initialize the `editButton` variable and then assign it using a series of if statememts:
```
let editButton
if (this.props.isEditable) {
   if (this.props.showInLine) {
     ...
   }
   else if (this.props.formFields) {
      editButton = ...
   } else {
      editButton = ...
   }
}
```

File: ui/shared/components/panel/view-fields/BaseFieldView.jsx [hunk 6]
```diff
@@ -113,6 +165,7 @@ BaseFieldView.propTypes = {
   showErrorPanel: PropTypes.bool,
   modalId: PropTypes.string,
   modalSize: PropTypes.string,
+  confirmDialog: PropTypes.string,
```
Comment (hanars) [inline ui/shared/components/panel/view-fields/BaseFieldView.jsx]: why did you add this prop?

File: ui/shared/components/panel/view-fields/BaseFieldView.jsx [hunk 7]
```diff
@@ -18,76 +20,126 @@ const FieldValue = styled.div`
...
-      ]}
-      {
-        hasValue && !props.hideValue &&
-        <FieldValue compact={props.compact} fieldName={props.fieldName}>
-          {props.fieldDisplay(fieldValue, props.compact, fieldId)}
-        </FieldValue>
-      }
-      {!props.fieldName && buttons}
-    </span>)
+  shouldComponentUpdate(nextProps, nextState) {
```
Comment (hanars) [inline ui/shared/components/panel/view-fields/BaseFieldView.jsx]: you should use `React.PureComponent` instead of `React.Component` and then you won't need this function at all

File: ui/shared/components/panel/view-fields/BaseFieldView.jsx (hunk 4 above)
Comment (lizzij) [reply ui/shared/components/panel/view-fields/BaseFieldView.jsx]: ESLint: The no-nested-ternary rule disallows nested ternary expressions.

File: ui/shared/components/form/ReduxFormWrapper.jsx [hunk 8]
```diff
@@ -268,7 +271,11 @@ class ReduxFormWrapper extends React.Component {
   }
 
   componentWillUpdate(nextProps) {
-    if (nextProps.submitSucceeded && nextProps.closeOnSuccess && !nextProps.noModal) {
+    if (this.props.onSubmitSucceeded && nextProps.submitSucceeded && nextProps.closeOnSuccess && nextProps.noModal) {
```
Comment (hanars) [inline ui/shared/components/form/ReduxFormWrapper.jsx]: I think if you pass an `onSubmitSucceeded` callback and submit succeeded you should call the callback but not call `handleClose` at all. So I don't think you need to check `closeOnSuccess` or `noModal` here

File: ui/shared/components/form/ReduxFormWrapper.jsx [hunk 9]
```diff
@@ -268,7 +271,11 @@ class ReduxFormWrapper extends React.Component {
   }
 
   componentWillUpdate(nextProps) {
-    if (nextProps.submitSucceeded && nextProps.closeOnSuccess && !nextProps.noModal) {
+    if (this.props.onSubmitSucceeded && nextProps.submitSucceeded) {
+      this.props.onSubmitSucceeded()
+      this.props.handleClose(true)
```
Comment (hanars) [inline ui/shared/components/form/ReduxFormWrapper.jsx]: I don't think this should call `handleClose` here. I think you should make this its own `if` statement separate from the `if/else` block that already existed so that way it does 2 separate checks, one for if it should call `onSubmitSucceeded` and the other for if it should call `handleClose`

File: ui/shared/components/form/ReduxFormWrapper.jsx [hunk 10]
```diff
@@ -319,14 +325,14 @@ const mapStateToProps = (state, ownProps) => ({
 })
 
 const mapDispatchToProps = (dispatch, ownProps) => {
-  return {
+  return (ownProps.onSubmitSucceeded ? null : {
```
Comment (hanars) [inline ui/shared/components/form/ReduxFormWrapper.jsx]: this should be unnecessary, since you just won't call these functions in the case with `onSubmitSucceeded`. Also, in theory, you should be able to support both a success callback and a modal, so I would take this out

File: ui/shared/components/panel/variants/FamilyVariantTags.jsx [hunk 11]
```diff
@@ -262,6 +262,7 @@ const FamilyVariantTags = (
           <NoteContainer>
             {savedVariant && savedVariant.notes.map(note =>
               <VariantNoteField
+                showInLine
```
Comment (hanars) [inline ui/shared/components/panel/variants/FamilyVariantTags.jsx]: I would actually move this prop directly into `VariantNoteField`, since you always want variant notes to have this prop

File: ui/shared/components/form/ReduxFormWrapper.jsx (hunk 10 above)
Comment (lizzij) [reply ui/shared/components/form/ReduxFormWrapper.jsx]: I might be wrong but I couldn't get the cancel button in `ButtonPanel` to work without this

File: ui/shared/components/panel/variants/FamilyVariantTags.jsx (hunk 11 above)
Comment (hanars) [reply ui/shared/components/panel/variants/FamilyVariantTags.jsx]: Try passing the `noModal` prop as well

File: ui/shared/components/panel/variants/FamilyVariantTags.jsx [hunk 12]
```diff
@@ -182,9 +182,11 @@ VariantTagField.propTypes = {
   family: PropTypes.object.isRequired,
 }
 
-const VariantNoteField = ({ action, note, variant, family, ...props }) => {
+const VariantNoteField = ({ action, note, variant, family, showInLine, noModal, ...props }) => {
```
Comment (hanars) [inline ui/shared/components/panel/variants/FamilyVariantTags.jsx]: you aren't using `showInLine` and `noModal` so theres no reason to declare them here

File: ui/shared/components/form/ReduxFormWrapper.jsx (hunk 10 above)
Comment (hanars) [reply ui/shared/components/form/ReduxFormWrapper.jsx]: with `noModal` you shouldn't need this

Comment (lizzij) [reply ui/shared/components/form/ReduxFormWrapper.jsx]: on submit/cancel: `TypeError: Cannot read property 'confirmOnClose' of undefined`

Comment (hanars) [reply ui/shared/components/form/ReduxFormWrapper.jsx]: That error is because `handleClose` is being called but it shouldn't be when `noModal` is passed. You need to figure out why that is happening and fix that issue, not get rid of the function entirely

File: ui/shared/components/form/RichTextEditor.jsx [hunk 13]
```diff
@@ -92,7 +92,7 @@ class RichTextEditor extends React.Component {
             }}
           />
         </div>
-        <div style={{ border: '1px #DDD solid', padding: '10px' }}>
+        <div style={{ width: '590px', border: '1px #DDD solid', padding: '10px' }}>
```
Comment (hanars) [inline ui/shared/components/form/RichTextEditor.jsx]: this should be `min-width`
